QUAL Send API response data before completing post-processing tasks

msoula opened the pull request:
This commit allows Rester to immediately send response data back to the client before completing some time consuming post-processing tasks (such as sending emails) in a termination callback if defined.

Termination callback should be created into the same class as the API method it terminates. Callback name must match the following format : `"_terminate_" + <API_METHOD_NAME> + "_" + <API_METHOD_RESPONSE_FORMAT_EXTENSION>`.

For instance, termination callback for orders _post_ method defined in `htdocs/commande/class/api_orders.class.php` must be named __terminate_post_json_.

Commit:
QUAL Sending API response data to clients before completing post-processing tasks.

This commit allows Rester to immediately send response data back to the client before completing some time consuming post-processing tasks (such as sending emails) in a termination callback if defined.

Termination callback should be created into the same class as the API method it terminates. Callback name must match the following format : "_terminate_" + <API_METHOD_NAME> + "_" + <API_METHOD_RESPONSE_FORMAT_EXTENSION>.

For instance, termination callback for orders "post" method defined in htdocs/commande/class/api_orders.class.php must be named "_terminate_post_json".

Commit:
Update index.php

Fix PHPScan error

Commit:
Add terminate callback

Add a termination function that will call termination API callback if existing.

Commit:
Fix copy error

eldy commented on htdocs/includes/restler/framework/Luracast/Restler/Restler.php:
```diff
@@ -1686,4 +1686,27 @@ protected function postCall()
             ), $this->responseData);
         }
     }
+	
```
Restler is an external library, so not developped by the Dolibarr team.
So to add new features in it, you should first try to post the PR to the upstream project...

eldy commented on htdocs/api/index.php:
```diff
@@ -434,4 +434,17 @@
 
+// Now flush output buffers so that response data is sent to request client
+ob_end_flush();
+ob_flush();
```
If we have a ob_end_flush before, this ob_flush should be useless, shouldn't it ?

msoula replied:
Agreed

eldy commented on htdocs/api/index.php:
```diff
@@ -434,4 +434,17 @@
+// Now flush output buffers so that response data is sent to request client
+ob_end_flush();
+ob_flush();
+flush();
```
We did a ob_end_flush, so it means we have returned the output to the client and we now have time to do all the rest. So why doing a flush here. It forces to make IO when they can be delayed and managed by the system when he can.

msoula replied:
Agreed

Commit:
Revert changes

Removes termination callback

Commit:
Move termination callback back to api/index.php

As discussed in the pull request conversation : 
- remove unnecessary flush calls ;
- remove termination callback method from Restler code and set it back into api/index.php

msoula commented:
I have made changes to the pull request in response to your comments. Thank you for your feedback.

Commit:
Fix indentation.

Commit:
Fix  PHPStan / php-stan check failure
